packages/babel-core/src/transformation/internal-plugins/shadow-functions.js

```diff
@@ -92,7 +92,10 @@ function remap(path, key) {
 
   fnPath.setData(key, id);
 
-  if (key === "this" && fnPath.isMethod({kind: "constructor"})) {
+  let classPath = fnPath.findParent((p) => p.isClassDeclaration());
```

Comment:
Doh, one last this, this should be `.isClass` since we want to cover `ClassExpression` too.